clouddriver-appengine/src/main/groovy/com/netflix/spinnaker/clouddriver/appengine/deploy/ops/TerminateAppEngineInstancesAtomicOperation.groovy

```diff
@@ -0,0 +1,81 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.appengine.deploy.ops
+
+import com.netflix.spinnaker.clouddriver.appengine.deploy.description.TerminateAppEngineInstancesDescription
+import com.netflix.spinnaker.clouddriver.appengine.deploy.exception.AppEngineOperationException
+import com.netflix.spinnaker.clouddriver.appengine.deploy.exception.AppEngineResourceNotFoundException
+import com.netflix.spinnaker.clouddriver.appengine.model.AppEngineInstance
+import com.netflix.spinnaker.clouddriver.appengine.provider.view.AppEngineInstanceProvider
+import com.netflix.spinnaker.clouddriver.data.task.Task
+import com.netflix.spinnaker.clouddriver.data.task.TaskRepository
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation
+import groovy.util.logging.Slf4j
+import org.springframework.beans.factory.annotation.Autowired
+
+@Slf4j
+class TerminateAppEngineInstancesAtomicOperation implements AtomicOperation<Void> {
+  private static final String BASE_PHASE = "TERMINATE_INSTANCES"
+
+  private static Task getTask() {
+    TaskRepository.threadLocalTask.get()
+  }
+
+  private final TerminateAppEngineInstancesDescription description
+
+  @Autowired
+  AppEngineInstanceProvider appEngineInstanceProvider
+
+  TerminateAppEngineInstancesAtomicOperation(TerminateAppEngineInstancesDescription description) {
+    this.description = description
+  }
+  /**
+   * curl -X POST -H "Content-Type: application/json" -d '[ { "terminateInstances": { "instanceIds": ["instance-1"], "credentials": "my-appengine-account" }} ]' localhost:7002/appengine/ops
+   */
+  @Override
+  Void operate(List priorOutputs) {
+    task.updateStatus BASE_PHASE, "Initializing termination of instances (${description.instanceIds.join(", ")}) in " +
+      "${description.credentials.region}..."
+
+    def credentials = description.credentials
+    def appengine = credentials.appengine
+    def accountName = credentials.name
+    def project = credentials.project
+
+    def instanceIds = description.instanceIds
+    def instances = instanceIds.collect { appEngineInstanceProvider.getInstance(accountName, credentials.region, it) }
+
+    instances.eachWithIndex { AppEngineInstance instance, index ->
+      def loadBalancerName = instance?.loadBalancers?.getAt(0)
+      def serverGroupName = instance?.serverGroup
+
+      if (!(instance && loadBalancerName && serverGroupName)) {
+        throw new AppEngineResourceNotFoundException("Could not find instance ${instanceIds[index]}")
```

Comment:
I wonder if it might be more user-friendly to attempt to resolve the required attributes of _all_ the instances that are going to be terminated, before attempting to terminate any of them. And then throw an exception if _any_ of them are missing (including details about everything that is missing). Just a thought.